### ISSUE TITLE
Routable content

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/OrchardCore#5305]
Dynamic Route Part: `/archives/{year}/{month}` such that any request matching this template would load this content item.

Then from a Liquid Part attached to the same content type we could use this:
`RouteData["year"]` and pass it to a query

Be able to pass default values in segments.

**Comment by sebastienros** [comment]
Could work on AutoroutePart with small changes probably.

**Comment by sfmskywalker** [comment]
Some quick thoughts and considerations:

* I'm thinking about updating the Autoroute feature to add support for matching against paths like `archive/{year}/{month}`, like ASP.NET Core routing does.
* I'm still figuring out how AutoroutePart works with `AutoRouteValuesAddressScheme` and `RoutePatternFactory` and if that's something that could be used for the pattern matching part.

That's it for now. If anyone has already more developd ideas about this implementation, I'd be all ears @sebastienros @deanmarcussen @jtkech @anyone :)

**Comment by jtkech** [comment]
@sfmskywalker no problem i will be able to help you as i did the route transformers and address schemes when migrating to 3.0, and @deanmarcussen added the support of contained item routing.

I will just have to re-put my mind on it

Do you think that we could do this feature after rc2? because in my `distributedCache` branch i already updated many services to be distributed aware, including autoroute entries and workflow route entries

**Comment by sfmskywalker** [comment]
That would be much appreciated, thank you JT. Sure let’s do it when you have time and space, I’m not in a hurry with this feature. I have a small list of things I want to work on, this just happened to be one of them and @sebastienros ’s favorite ;)

I’ll patiently (and eagerly) wait for you.

**Comment by deanmarcussen** [comment]
Cool, be great to have someone work on this.

From some initial investigation some time ago (so from what I remember, and before I forget again)

The `AutorouteTransformer` is where it resolves a route.
The `AutorouteValuesAddressScheme` is where the `UrlHelper` or `LinkGenerator` generates a pretty route.

One of the difficulties in potentially reusing some of the ASP.NET Core functionality is that a lot of the DFA Matchers and other classes have become internal.

Also an ASP.NET routing pattern may not be the most appropriate, because we also use `Liquid` patterns in `Autoroute`, so {route?} values have the possibility of conflicting with liquid.

When working with the `DynamicRouteValueTransformer` it is important that only one `Transformer` (or other route) returns a result, or you get `AmbiguousMatchExceptions` and the site comes down because the routing is broken :(

My original idea for this, was to be able to work with the `AutorouteTransformer` but split the path it is matching on (to keep it simple)

- `archive/yy/mm/dd` matches on
- `archive`

and owns the entire `archive` path (i.e. no other autoroute can have `archive/i'm-another-route`)

then to place all the segments after the matched path into the route dictionary, (either by name if the pattern defines a name, (so `year`), or if they extend past the pattern, as index values, i.e. `minute_3`

Then you could pull these values back in liquid via your `Route` pr

So it maybe that the pattern matching is a lot simpler than ASP.NET Core, and has less features, but can be calculated as a path-pattern set of segments when building and mutating the `AutorouteEntries` 

Then the lookup would have to iterate path segments more, to find a match. Rather than matching on the entire path, as it does currently.

@jtkech may have some better ideas too :)

**Comment by sfmskywalker** [comment]
@deanmarcussen Brilliant, that's very helpful. Then I'll continue the path of doing the pattern matching myself. Thanks!

**Comment by jtkech** [comment]
> @jtkech may have some better ideas too :)

Maybe not yet or nothing more, so many other things to think about ;)

Hmm, in place of a new `RoutePart` maybe extend the `AutoroutePart` and its settings UI where we could add (or not) multiple key / value pairs route data, e.g. with liquid support, then allow to enter a path as before but also with `/{segmentKey1}/{segmentKey2}` as a route pattern.

Then when updating an item, the autoroute part could evaluate its route values and its path by replacing the segments with the related values. Here it's not a pattern matching but a path building from a pattern and at the edit / update time, because at the end an item has only one path.

Then nothing to change in the address scheme, and in the transformer, as we already add the `contentItemId` route value, we could add the ones provided by the autoroute part, but here we don't have the content item, we could get it and then it will be cached for the `ItemController`, but maybe better to do it in the controller, just tried it works.

            HttpContext.Request.RouteValues.TryAdd("routeKey", "routeValue");

Or without changing the `ItemController`, maybe better that the driver of this new `AutoroutePart` would add these route values on displaying. This way nothing to change in the address scheme, autoroute transformer and `AutorouteEntries`. Just implement the settings UI to input multiple route values with liquid support, and a path pattern that will use these route values on updating.

**Comment by sfmskywalker** [comment]
@jtkech Agreed, extending AutoroutePart makes sense. 

> Then when updating an item, the autoroute part could evaluate its route values and its path by replacing the segments with the related values.

Perhaps I misunderstood, but it sounds like that would cause the `AutoroutePart` to store a path with the route placeholders (e.g. `{segmentKey1}` replaced with the provided default values (if any).

For example, let's say I am configuring a route like so:

`/archive/{year}`

And I provide default values for {year} and {month}, maybe using liquid like {{ DateTime.Now.Year }}.
When I now save a content item with an `AutoroutePart` with that route template, the path would be saved as `/archive/2020`. 

But that's not what we want. The route of the content item should still be `/archive/{year}` when updating the content item.

Only when an HTTP request comes in for e.g. `/archive`, should the missing `{year}` segment be replaced with the configured default value with `2020` (the result of the liquid-evaluated default value for the `{year}` segment as configured on the `AutoroutePart` or its content type settings).

But perhaps I misunderstood what you were saying.

**Comment by jtkech** [comment]
@sfmskywalker 

- Okay, i thought that `/archive/2019` and `/archive/2020` would be related to 2 different content items. With your option the url would be `/archive` and the related item part will provide route values. But maybe @deanmarcussen understood it better, maybe this is the request with the generated url `/archive/2019` or `/archive/2020`, still related to the same content item, that provides the values. In that case the autoroute part settings would just need to provide a `Pattern` using the keys of the route data that may be passed (maybe optional), not their values.

    It would be like in a razor page with `@page "/archive/{year}"` where year could be optional `{year?}`. If so, we could use directly e.g. `/archive/2020` or an url helper by providing the route values including the year or the `ContentLinkTagHelper` like this

      <a display-for="@contentItem" asp-route-year="2020">@contentItem.DisplayText</a>

- The `AutorouteEntry` would need a `Pattern` property e.g. `/archive/{year}/{month}`,  maybe the `/archive` segment is not needed here as it is resolved as before through the entry path that would still be `/archive`, but maybe better to include it to build an endpoint (see below).

- Then in `AutoRouteValuesAddressScheme` we would use this `Pattern` to retrieve the route keys, then retrieve the values from the `address.ExplicitValues` e.g. if the link tag helper was used or another url helper. And update the route values to build the `RouteEndpoint` by using

      RoutePatternFactory.Parse(autorouteEntry.Pattern, routeValues, null),

    Here maybe combine the entry `Path` with the `Pattern` if it doesn't include the root `Path`

- Then in `AutoRouteTransformer`, as suggested by @deanmarcussen, if no entry is found maybe just check that the current path starts with one of the entry paths (limited to the entries that have a `Pattern`). Then use the entry `Pattern` to retrieve the route keys and use the current path to retrieve their values, and then update the route values that we return. Hmm, if there are missing values maybe we could provide default values e.g. `{year=2020}`.

    Maybe above and here we could use some aspnetcore classes as `RoutePattern` or just `RouteTemplate`, `RoutePatternMatcher` or just `TemplateMatcher`. E.g. see [this unit test](https://github.com/dotnet/aspnetcore/blob/master/src/Http/Routing/test/UnitTests/Template/TemplateMatcherTests.cs#L20)

**Comment by sfmskywalker** [comment]
@jtkech 

That is most valuable input, thank you! I think with all of the above I have enough to start with.

**Comment by deanmarcussen** [comment]
@agriffard yes, you could run a sql query / lucene query with this feature. but this is the issue we need @sfmskywalker to work on to do that ;)
